### ISSUE TITLE
[Improvement]: Implement FOR UPDATE in Concrete\Dao::getData()

### ISSUE THREAD

**Issue by solverat** [pimcore/pimcore#17781]
### Improvement description

We experienced some ugly race conditions within a headless application. Concurrencies occurs when multiple requests (from client API and also from server API) are getting processed at almost the exact same time.

I was able to reproduce it locally and located the issue here:

https://github.com/pimcore/pimcore/blob/322366f2d0d7613a0ce7ac796092f733b165008f/models/DataObject/Concrete/Dao.php#L135-L137

After changing the SQL statement to 

```sql
SELECT * FROM object_store_' . $this->model->getClassId() . ' WHERE oo_id = ? FOR UPDATE
```

...the issue was gone.

Of course, I'm able to provide a PR but I want some feedback first. `FOR UPDATE` has been already added [several times](https://github.com/search?q=repo%3Apimcore%2Fpimcore%20%22FOR%20UPDATE%22&type=code) and also MySQL does not throw any exception if we're going to add this - so I guess it would be fine.

However, if you're **not** willing to implement this, I could override the DAO model, but then I would have to clone the whole method (`getData()`, which is something I want to prevent. Moving those three lines to a dedicated method (like `getRawData`) would be great.

**Comment by brusch** [comment]
@solverat thanks for sharing your finding - in theory there could be a bad performance impact by adding `FOR UPDATE`, because this method is also called for normal read operations. 
In reality this probably doesn't happen that frequent, depending on the update frequency and the cache config. 

But I'm wondering if you also experience race-conditions in combination with the cache, because I'd rather experience them there? 🤔

**Comment by solverat** [comment]
@brusch I disabled the cache completely to test this case. Cache was also my first guess, but that's something we could fix within the application. In fact, disabling the cache was the reason I stumbled on that. Because in my test case, I fetched every entity by `[force => true]`.

Regarding performance: True (I didn't experience any bad performances after changing it, but I haven't run any benchmarks either. :) ).

Two options, if you don't want to change it: 
- Configurable via class definition (Checkbox `Lock read transactions` for example)
- Just extract the data SQL query into a new method so a custom DAO class can easily override it

**Comment by brusch** [comment]
According to https://mariadb.com/kb/en/for-update/  I'd assume it shouldn't have a (noticable) bad impact on performance, because when just reading the object from database, it's not enclosed by a transaction, therefore `FOR UPDATE` is ignored and no locking takes place - also locking is on row level. 

> The FOR UPDATE clause of [SELECT](https://mariadb.com/kb/en/select/) applies only when [autocommit](https://mariadb.com/kb/en/server-system-variables/#autocommit) is set to 0 or the SELECT is enclosed in a transaction. A lock is acquired on the rows, and other transactions are prevented from writing the rows, acquire locks, and from reading them

So I'd go for adding `FOR UPDATE`. 
Can you create a PR for us? 

Thanks a lot!

**Comment by brusch** [comment]
Fixed by #17788

**Comment by solverat** [comment]
🥳 Awesome! Thanks, @brusch!

**Comment by Jan-Walther-Basilicom** [comment]
Hi @solverat, hi @brusch,

Can this change cause deadlocks when 2 processes try to load the same object at the same time (let's ignore Redis cache)? I face some of them with parallel imports when 2 processes try to load the same data object.

- first process loads the object -> row gets locked until the object gets saved or the PHP process finishes
- in parallel, second process tries to load the same data object but will get a `Lock wait timeout exceeded; try restarting transaction` error (if the first process is slower than [innodb_lock_wait_timeout](https://dev.mysql.com/doc/refman/8.4/en/innodb-parameters.html#sysvar_innodb_lock_wait_timeout))

**Comment by dpfaffenbauer** [comment]
@Jan-Walther-Blackbit yes, thats exactly what FOR UPDATE does. it locks other FOR UPDATE select queries

**Comment by Jan-Walther-Basilicom** [comment]
Hmm, I do not understand why we want to lock objects when only reading data. @solverat, can you explain more about the race condition which you saw? Does this happen when 2 API requests want to write data for the same object? If yes, then wouldn't it make more sense to use
https://github.com/pimcore/pimcore/blob/514e9ceaa22820280c6fc0e8a66fa9beef9d95e2/models/DataObject/AbstractObject/Dao.php#L635-L642
before saving?

Alternative solution: According to [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html) locking with `FOR SHARE` seems to be what we are looking for:

> SELECT ... FOR SHARE 
> Sets a shared mode lock on any rows that are read. Other sessions can read the rows, but cannot modify them until your transaction commits. If any of these rows were changed by another transaction that has not yet committed, your query waits until that transaction ends and then uses the latest values. 

> SELECT ... FOR UPDATE
> For index records the search encounters, locks the rows and any associated index entries, the same as if you issued an UPDATE statement for those rows. Other transactions are blocked from updating those rows, from doing SELECT ... FOR SHARE, or **from reading the data in certain transaction isolation levels**.

**Comment by solverat** [comment]
The `FOR UPDATE` lock is intentional and solves a specific race condition we encountered in a payment process:

Two nearly simultaneous requests - both loaded the object with a given status at the (almost exact) same time, so both executed a workflow transition which then executed several tasks twice.

With `FOR UPDATE` on load:

- First request locks the row, executes workflow, saves with new status, releases lock.
- Second request waits, then loads the object with an updated status and so on.

`FOR SHARE` only prevents writes while reading data, but it doesn't prevent a second process from also reading the same (outdated) data in parallel.
`FOR UPDATE` is necessary because it ensures the second process waits until the first one has completed its transaction.

Regarding your import scenario:
Yes, parallel processes loading the same object will have to wait for each other - but this is intentional. If both processes modify and save the object, one needs to wait for the other (transaction) to finish, otherwise we get the same race condition (both work with outdated data and overwrite each other's changes).

The lock wait timeout only triggers if a process takes extremely long (> `innodb_lock_wait_timeout`, usually 50 seconds). If you're experiencing lock wait timeouts during imports, you may want to check if your import process keeps transactions open for too long.

**Comment by IronSean** [comment]
@Jan-Walther-Blackbit I haven't seen you since Pimcore Inspire a couple years ago, it's nice to see you have similar thoughts to me.

@solverat. I do not agree that the approach here is the universally correct one. I think for your use case there were other alternatives which could opt into a locking behavior to solve your issue. For instance, in pre and post listeners, or in the controller endpoints you could have used a symfony lock backed by the database or redis to prevent multiple operations reading the same object at the same time. 

However, with this current implementation of the Dao::getData() method forcing a lock for everything that is loaded it causes everything to be locked even if it is not intended to be written. Especially, this matters not just for the root object being loaded but any related objects which are being loaded. 

For example a save listener for a Product object which loads the assigned Manufacturer object in a many to one relation to check apply logic: There is no need for a write lock as that Manufacturer object is not being modified, but this prevents any other Products from continuing their save on other threads. 

Under normal operation this may not be an issue, but imagine the case where you are importing 500,000 products from a single Manufacturer because they gave you an update file. Now all 500,000 rows must get a lock on the same Manufacturer object in the save listener because they are all guaranteed to be referencing the same object. This turns what could be a fully parallel import to one which is much more sequential. 

Other possible solutions which would not have this issue:
1. Implement the read/write locks which are needed in the workflows that need them (via locks in the controller or listeners relevant).
2. Add a parameter to the `getX()` methods which map back to the Dao::getData() method, which allows you to pass a flag for a database lock when needed.

In my specific case removing the FOR UPDATE statement has increased throughput by 50% on an operation where I'm re-saving all the items which point to the same related object, which is going to save about 100 hours at the current pace.

**Comment by dpfaffenbauer** [comment]
@IronSean We tried handling it in certain events with all kinds of locks and nothing worked, requests came in too fast for something like this too work. we had to find a solution were the database takes care about this

**Comment by kanevbg** [comment]
I agree with @IronSean , locking should not be forced in abstractions but in the workflow specific implementations trough friendly API.
EG: https://www.doctrine-project.org/projects/doctrine-orm/en/3.5/reference/transactions-and-concurrency.html#locking-support

**Comment by IronSean** [comment]
Thanks for the extra context @dpfaffenbauer, I've had good luck with the Symfony Lock component, but I could see where in high volume operations the overhead would become unmanageable and a database row-level lock is much more efficient. 

Thinking on it overnight, I've been considering two main approaches to this:

### Option 1: Make the FOR UPDATE optional via the Pimcore APIs.

Update the method signature to something like `public function getData(?bool $writeLock = false): void` where the community could decide the preferred default behaviour but it could be overridden when optimization requires it. I think this would involve updating the signature of subclassed core library files, and updating the generation code for the DataObject class generation to expose this flag on all relevant methods (for instance the `getMyRelation()` generated method. There is also consideration in how the Listing classes work by default.

This looks like a cleaner solution to me, but I haven't fully digested the scope of the change required yet.

### Option 2: Try to refactor my current projects affected by this.

I think I have a worst case example of how this can negatively impact performance because I have a dataset of 3m Products of which 90% reference the same object via relation. Actually it is a relation within a field collection to support having extra metadata on the relation with more structure than the Advanced Many-to-Many provides. 

In my case we have calculated fields which load these relations in order to create an easy grid-filterable value for the client to filter for Products only related to relation A, or only related to relation B. That calculated field has to load Relation A and B with every save operation which the database locking prevents happening in parallel. In this case I know it is a read-only operation I am doing. 

Some alternatives I see:
1. Update to Studio UI and use the generic index and PQL instead of the calculated field: this will work, though it will be slightly less user friendly as the end users would be looking at PQL to filter via field collection then relation.
2. Replace my use of Pimcore APIs with direct database queries where I can control it as a read-only operation: This is possible, though I prefer to stay within the Pimcore APIs when possible to avoid breaking changes if the underlying tables were to be updated. This also would not automatically cache in memory the first time it is loaded the same way it would with the Pimcore APIs, and lead to either redundant queries or needing to manually cache the result.
3. Move away from calculated fields to read-only fields and a save listener which can optimize it so these fields are only re-calculated when necessary: in my case these relations are managed and syndicated from an external system, so only during those systems updates would I need to recalculate the helper values. This would avoid the performance issue for other operations, but it would remain for syndications from that system.

**Comment by Externaluse** [comment]
# Suggestion for a fix / @brusch Request to reopen

## Current behaviour ##

Since Pimcore 11 with the original PR here presented, the object hydration path uses `SELECT … FOR UPDATE` in several places where Pimcore 10 previously performed plain reads.  
In normal UI usage this is reasonable and helps consistency.

However, during **parallel batch imports** (e.g. Data Director, custom CLI importers) this change can lead to frequent deadlocks — not during the primary object save itself, but while persisting the **version snapshot**.

Observed call chain where deadlocks materialize:
```
Concrete->saveVersion()
  Version->save()
    Element\Service::loadAllFields()
      DataObject\Service::loadAllObjectFields()
        AbstractObject::getById()
          Concrete\Dao->getData()
```
These reads now use` FOR UPDATE`.
With larger, relation-heavy objects (Many-To-Many, classifications, etc.) this cascades into additional locked reads across the object graph.

A typical cycle:
```
Process A locks Object 100
Process B locks Object 200
Process A needs 200
Process B needs 100
→ deadlock
```

The final exception often surfaces at:
`DELETE FROM edit_lock WHERE cid = ?`
which itself is harmless, but becomes part of the same lock graph and closes the cycle.

### Why this shows up mostly in versioning
Version snapshotting is read-heavy, not mutation-heavy, but it currently hydrates the entire object graph with write-intention locks.

For UI edits this is fine.
For automated parallel imports, this is often unnecessarily strict.

### Files Involved (11.5) in the suggestion

[models/DataObject/AbstractObject.php](https://github.com/pimcore/pimcore/blob/11.5/models/DataObject/AbstractObject.php)

[models/DataObject/Concrete/Dao.php](https://github.com/pimcore/pimcore/blob/11.5/models/DataObject/Concrete/Dao.php)

The relevant change is the unconditional addition of `FOR UPDATE` in DAO reads and inheritance hydration.

### Proposal: Opt-In Read-Only Hydration Mode for Version Snapshots

Instead of changing default behavior, introduce an opt-in context flag that allows integrating systems (batch importers, migration tools, etc.) to signal:

> “This hydration is for version snapshotting only; strict pessimistic locking is not required.”

### Rationale

- Keeps current default semantics for UI / interactive edits.
- Allows high-volume importers to reduce lock contention.
- Avoids API signature changes.
- Similar in spirit to how Doctrine exposes different lock modes rather than forcing one globally.

## Suggested Implementation
1. AbstractObject – introduce static context flag
```
// models/DataObject/AbstractObject.php

/**
 * @internal
 * When enabled, object hydration is read-only (no FOR UPDATE).
 * Intended for version snapshotting in batch contexts to reduce deadlocks.
 */
protected static bool $versionSnapshotReadOnly = false;

public static function getVersionSnapshotReadOnly(): bool
{
    return self::$versionSnapshotReadOnly;
}

public static function setVersionSnapshotReadOnly(bool $flag): void
{
    self::$versionSnapshotReadOnly = $flag;
}

public static function doVersionSnapshotReadOnly(): bool
{
    return self::$versionSnapshotReadOnly;
}
```
Pattern follows existing static flags like hideUnpublished.

2. `Concrete\Dao::getData()` – conditional locking
```
$sql = 'SELECT * FROM object_store_' . $this->model->getClassId() . ' WHERE oo_id = ?';

if (!AbstractObject::doVersionSnapshotReadOnly()) {
    $sql .= ' FOR UPDATE';
}
```
3. Inheritance hydration (object_query_*), incl. update to the existing commment
```
// When hydrating for version snapshot only, strict row locking may be unnecessary.
// Integrators can enable read-only hydration via AbstractObject::doVersionSnapshotReadOnly()
// to reduce lock contention during automated or parallel imports.
// This trades strict parent consistency for lower deadlock probability.
$sql = 'SELECT * FROM object_query_' . $this->model->getClassId() . ' WHERE oo_id = ?';

if (!AbstractObject::doVersionSnapshotReadOnly()) {
    $sql .= ' FOR UPDATE';
}
```
4. Integrator use example (whereby du to self:: AbstractObject could be replaced with $object::...):
```
$old = AbstractObject::getVersionSnapshotReadOnly();
AbstractObject::setVersionSnapshotReadOnly(true);

try {
    $object->saveVersion(false, false, $note);
} finally {
    AbstractObject::setVersionSnapshotReadOnly($old);
}
// Important: restore via finally to avoid leaking state in long-running workers.
```
## Alternatives Considered
1. Plain SELECT
Removes most deadlocks but weakens consistency globally.
2. FOR SHARE
Reduces impact but still contributes to lock graphs and is not universally supported.
3. Function Signature Changes (e.g. passing lock mode)
Technically clean, but touches many call sites and public APIs.
A static context flag is closer to existing Pimcore patterns and less invasive.

## Summary
- Default behavior remains unchanged.
- Integrators can use simple testing for the existence of the new methods in version compatibility instead of needing to test for the availability of parameters; no signature changes anywhere
- Batch importers gain a safe opt-in escape hatch.
- No signature changes.
- Deadlocks in relation-heavy parallel imports are significantly reduced in testing.
- Versioning becomes much more light-weight again if opting in

A PR with the above changes can be provided if the direction is acceptable. The patch is presently implemented in my environment and dropped DeadlockExceptions from ~8k/day to almost none.

**Comment by Externaluse** [comment]
@brusch Sorry to bother; is there any chance that reopening this might be considered based on the suggestion, or should I file a new issue referring to this one?

**Comment by pdchaudhary** [comment]
I agree with this @Externaluse . The unconditional SELECT ... FOR UPDATE during hydration increases lock contention and creates predictable deadlock cycles in parallel import scenarios, especially with relation-heavy graphs during version snapshotting. An opt-in read-only hydration mode is a reasonable and minimally invasive mitigation.

**Comment by brusch** [comment]
@Externaluse since this was a PR, it would make sense to create a new issue/PR for that. 
I see your points, I like the idea of keeping the current behavior as the default and introduce a flag for changing it, e.g.

```php
$object = DataObject\Foo::getById(123, ['lock' => false])
```

WDYT? 
Happy to merge a PR for that 😉

**Comment by Externaluse** [comment]
@brusch Thanks for the quick response! This would be my first contribution to the project, so I'll read through the
  guidelines carefully. I've only just returned to PHP and Pimcore in particular a little while ago; may I clarify a few things before I start an issue and PR?

  ### Mapping the `FOR UPDATE` landscape

There appear to be 4 `FOR UPDATE` sites in the DataObject Dao layer (v11.5.13):

  **1. `Concrete\Dao::getData()` — fires during `getById()` (hydration)**

https://github.com/pimcore/pimcore/blob/9246a42d914fc3f75c9d90bec7b946a561e7be6f/models/DataObject/Concrete/Dao.php#L132-L138

  **2. `Concrete\Dao::update()` — fires during `save()` (inheritance)**
  https://github.com/pimcore/pimcore/blob/v11.5.13/models/DataObject/Concrete/Dao.php#L280-L286

  **3. `Localizedfield\Dao::update()` — fires during `save()` (localized inheritance)**
  https://github.com/pimcore/pimcore/blob/v11.5.13/models/DataObject/Localizedfield/Dao.php#L338-L344

  **4. `Objectbrick\Data\Dao::update()` — fires during `save()` (brick inheritance)**
  https://github.com/pimcore/pimcore/blob/v11.5.13/models/DataObject/Objectbrick/Data/Dao.php#L150-L156

  Sites 2–4 all lock parent rows during inheritance resolution while *writing*. I'd assume those locks should stay
  unconditional — if you're in `save()`, you need them. I'm listing them here for completeness, but I believe the only
  site an integrator would ever want to opt out of is **site 1**: the `FOR UPDATE` during read-only hydration.

  Since site 1 fires inside `getById()` → `Dao::getData()`, both a static flag and an options param on `getById()` can
  reach it equally well.

  ### Trade-offs as I see them

  **Options param (`['lock' => false]`):**
  - Per-call intent is explicit — no global state, no risk of forgetting to restore
  - Clean, self-contained call sites

  **Static flag (following `$hideUnpublished` / `$getInheritedValues` / `$disableDirtyDetection`):**
  - Block-level convenience — one toggle wraps a whole batch of reads (e.g. a full import run) without having to pass
  the option on every `getById()` call
  - Easy feature detection for integrators: `method_exists(AbstractObject::class, 'setForUpdateEnabled')` vs having to
  deal with `OptionsResolver` throwing `UndefinedOptionsException` on unknown keys or through using reflection.

  ### A question on direction

  You suggested the options param approach — could I ask what your reasoning is for preferring that over an additonal static
  flag? I notice Pimcore already has three static flags on `AbstractObject` following the same save/restore pattern, so
  I want to understand whether there's a design direction to move away from those, or whether it's more about this
  specific case. Happy to go either way — just want to make sure the PR aligns with where you want the codebase to go. My personal preference is the static flag due to the feature detection issue.

  ### A note on caching

  I'm not sure if I see this entirely correctly, but I could imagine there's an interaction worth considering with the
  options param approach: if `DataObject\Foo::getById(123, ['lock' => false])` is called and the object is already in
  RuntimeCache from a previous `getById(123)`, the `lock` param would be silently ignored — the cached object is
  returned without hitting the database at all. The `force` param bypasses the cache, but that also forces a full
  reload. It may not be an issue in practice, but it felt worth flagging. With your experience, do I need to worry/consider this here?

  Looking forward to your thoughts — happy to start on the PR once we've aligned on the approach.
